docs/plugin_author_guide/compiler_plugins.rst

```diff
@@ -0,0 +1,237 @@
+the decorated function to return a special ``MLIRFunc`` object which contains
+the source code of the algorithm to be compiled with MLIR.
+
+.. _Numba compiler plugin: https://metagraph-numba.readthedocs.io
+.. _MLIR compiler plugin: https://metagraph-mlir.readthedocs.io
+
+Compilable concrete algorithms should avoid copying significant amounts of
+data to or from temporary containers when calling compiled code.  The type
+annotation for the function should indicate which concrete types the code can
+directly use, and Metagraph's automatic translation system should handle the
+translation.  That will ensure the translation step is visible to the user
+(and the optimizer) in the Dask task graph, rather than being hidden inside
+the implementation of the compiler plugin.
+
+
+Invoking the Compiler
+---------------------
+
+Users will generally not need to interact with or think about the compiler
+when using Metagraph.  The optimizer is applied automatically when a Metagraph
+placeholder object is computed.  
+
+If you have a larger DAG that uses Metagraph for an intermediate calculation,
+you will have to ask Dask to apply the Metagraph optimizer manually.  To do this:
+
+ .. code-block:: python
+
+    import dask
+    import metagraph as mg
+    # res is a Dask object with internal Metagraph tasks
+    res_opt = dask.optimize(res, optimizations=[mg.optimize])
+    answer = res_opt.compute()
+
+The Metagraph optimizer will leave all non-Metagraph tasks unchanged, so it is
+always safe to apply.
+
+
+Visualizing Compilation
+-----------------------
+
+Metagraph placeholder objects have a custom ``visualize()`` method which works
+the same as the standard `Dask visualize() method`_, but with special shapes
+and labels for Metagraph operations.  For example, this DAG:
+
+.. image:: visualize.png
+
+shows translation steps with ellipses, the concrete type of results with
+parallelograms, and algorithms with octagons.
+
+As with optimization, the custom Metagraph visualize method can be used with
+any Dask object by calling it directly:
+
+.. code-block:: python
+
+    import metagraph as mg
+
+    mg.visualize(my_dask_object)
+
+When the DAG contains compilable tasks, they will be highlighted with a single
+red octagon outline:
+
+.. image:: vis_unopt.png
+
+And when the optimizer has compiled and fused tasks, the tasks will be shown
+in a double octagon outline with a label listing the algorithms that were fused:
+
+.. image:: vis_opt.png
+
+By default, the visualizer optimizes the graph before drawing it.  To disable
+this, pass ``optimize_graph=False`` to the ``visualize()`` method.
+
+.. _dask visualize() method: https://docs.dask.org/en/latest/graphviz.html
+
+
+A Compiler Example
+------------------
+
+Consider the following example of task graph:
+
+.. image:: dag_orig.png
+
+The user who constructed this graph will have in their script a Metagraph
+placeholder object which contains a “future” representing the result of
```

Comment:
Since "placeholder" has been highlighted using backticks earlier in this document, it might be a good idea to use them here for consistency.

Reply:
It might be a good idea to have "future" link to https://en.wikipedia.org/wiki/Futures_and_promises since it's a somewhat commonly known term instead of having it in quotes. This will make it clear what's being referred to without using quotes. 

This is a stylistic suggestion, so I'll leave it up to you if you think it's a good idea to implement.

Reply:
Fixed.